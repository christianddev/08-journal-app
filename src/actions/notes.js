
import Swal from 'sweetalert2';
import {db} from '../firebase/firebase-config';
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from '../helpers/loadNotes';
import { types } from '../types/types';


export const activeNote = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note
  }
})

export const addNewNote = (id, note) => ({
  type: types.notesAddNote,
  payload:{
    id,
    ...note
  }
})

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const {auth: {uid}} = getState();
    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    }
    try {
      const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
      dispatch(activeNote(doc.id, newNote));
      dispatch(addNewNote(doc.id, newNote));
    } catch (error) {
      console.log('error', error)
    }
  }
}

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes
})

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  }
}

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const {auth: {uid}} = getState();
    if (!note.url) {
      delete note.url;
    }

    const noteToFirestore = {...note};
    delete noteToFirestore.id;
    try {
      await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore)
      dispatch(refreshNote(note.id, noteToFirestore));
      Swal.fire({
        title: 'Saved',
        text: note.title,
        icon: 'success'
      });
    } catch (error) {
      Swal.fire({
        title: error?.code,
        text: error?.message,
        icon: 'error'
      });
      console.error('error startSaveNote: ', error);
    }
  }
}

export const refreshNote = (id, note) => ({
  type: types.notesUpdated,
  payload: {
    id, note: {
      id,
      ...note
    }
  }
})

export const startUploading = (file) => {
  return async (dispatch, getState) => {
    const {notes: {active: note}} = getState();
    Swal.fire({
      title: 'Uploading',
      text: 'Please wait',
      allowOutsideClick: false,
      willOpen: () => {
        Swal.showLoading();
      }
    });
    const fileUrl = await fileUpload(file);

    Swal.close();

    note.url = fileUrl;
    dispatch(startSaveNote(note));

  }
}

export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: {
    id
  }
})

export const startDeleting = (id) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    try {
      await db.doc(`${uid}/journal/notes/${id}`).delete()
      dispatch(deleteNote(id));
      Swal.fire({
        title: 'Deleted',
        text: 'Picture',
        icon: 'success'
      });
    } catch (error) {
      Swal.fire({
        title: error?.code,
        text: error?.message,
        icon: 'error'
      });
      console.error('error startDeleting: ', error);
    }
  }
}

export const noteLogout = () => ({
    type: types.notesLogoutCleaning
});

