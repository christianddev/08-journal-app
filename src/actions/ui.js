import { types } from "../types/types"

export const serError = (error) => ({
  type: types.uiSetError,
  payload: error
})

export const removeError = () => ({
  type: types.uiRemoveSetError
})

export const startLoading= () => ({
  type: types.uiStartLoading
})

export const finishLoading = () => ({
  type: types.uiFinishLoading
})