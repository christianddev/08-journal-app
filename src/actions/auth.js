import Swal from 'sweetalert2';
import { types } from '../types/types';
import {firebase, googleAuthProvider} from '../firebase/firebase-config';
import { finishLoading, startLoading } from './ui';
import { noteLogout } from './notes';

export const login = (uid, displayName) => (
  {
    type: types.login,
    payload: {
      uid,
      displayName
    }
  }
)

export const logout = () => (
  {
    type: types.logout
  }
);

export const startLoginWithEmailPassword = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());
    firebase.auth().signInWithEmailAndPassword(email, password)
      .then(({user}) => {
        dispatch(finishLoading());
        dispatch(
          login(user.uid, user.displayName)
        );
      } )
      .catch((error) => {
        Swal.fire({
          title: error?.code,
          text: error?.message,
          icon: 'error'
        });
        console.error('error startLoginWithEmailPassword: ', error);
        dispatch(finishLoading());
      });
    }

}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return ( dispatch) => {
    firebase.auth().createUserWithEmailAndPassword(email, password)
      .then( async({user}) => {
        await user.updateProfile({displayName: name})
        dispatch(
          login(user.uid, user.displayName)
        );
      })
      .catch((error) => {
        Swal.fire({
          title: error?.code,
          text: error?.message,
          icon: 'error'
        });
        console.error('error startRegisterWithEmailPasswordName: ', error)
      }
        );
  }

}

export const startLoginGoogle = () => {
  return (dispatch) => {
    firebase.auth().signInWithPopup(googleAuthProvider)
      .then(({user}) => {
        dispatch(login(user.uid, user.displayName));
      })
  }
}

export const startLogout = () => {
  return async (dispatch) => {
    await firebase.auth().signOut();
    dispatch(logout());
    dispatch( noteLogout() );
  }
}