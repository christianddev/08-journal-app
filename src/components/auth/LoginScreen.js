import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from 'react-router-dom';
import { startLoginGoogle, startLoginWithEmailPassword } from "../../actions/auth";
import { useForm } from "../../hooks/userForm";
import validator from 'validator';
import {serError, removeError }  from '../../actions/ui';

const LoginScreen = () => {

  const dispatch = useDispatch();
  const { ui: { msgError, loading } } = useSelector(state => state);

  const [formValues, handleInputChange] = useForm({
    email: 'userxd1@email.com',
    password: '123123'
  })

  const isFormValid = () => {
    if (!validator.isEmail(email)) {
      dispatch(serError('Invalid email'));
      return false;
    }else if (password.length < 5) {
      dispatch(serError('Password should be at least 6 characters'));
      return false;
    }

    dispatch(removeError());
    return true;
  }

  const handleLogin = (event) => {
    event.preventDefault();
    if (isFormValid()) {
      dispatch(startLoginWithEmailPassword(email, password));
    }
  }

  const handleGoogleLogin = () => {
    dispatch(startLoginGoogle());
  }

  const {email, password} = formValues;

  return (
    <div>
      <h3 className="auth__title mb-5">Login</h3>
      <form onSubmit={handleLogin}>
        {
          msgError && ( <div className="auth__alert-error"> {msgError} </div> )
        }
        <input
          type="text"
          placeholder="Email"
          name="email"
          className="auth__input"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
           />
        <input
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
          autoComplete="off"
          value={password}
          onChange={handleInputChange}
           />
        <button
          className="btn btn-primary btn-block mb-5"
          type="submit"
          disabled={ loading }
          >Login</button>
        <hr />
        <div className="auth__social-networks">
          <p>Login with Social Networks</p>
          <div
            className="google-btn"
            onClick={handleGoogleLogin}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
          <Link
            className="link"
            to="/auth/register"
            >
            Register
          </Link>
        </div>
      </form>
    </div>
  );
};

export default LoginScreen;
