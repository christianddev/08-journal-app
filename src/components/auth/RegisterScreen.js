import React from 'react';
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/userForm';
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';
import {serError, removeError }  from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

const RegisterScreen = () => {

  const dispatch = useDispatch();
  const { ui: { msgError } } = useSelector(state => state);

  const [formValues, handleInputChange ]= useForm({
    name: 'userXD',
    email: 'userxd1@email.com',
    password: '123123',
    password2: '123'
  });

  const {name, email, password, password2} = formValues;

  const handleRegister = (event) => {
    event.preventDefault();
    if(isFormValid()) {
      dispatch(startRegisterWithEmailPasswordName(email, password, name));
    }
  }

  const isFormValid = () => {
    if (name.trim().length === 0) {
      dispatch(serError('Name is required'));
      return false;
    } else if (!validator.isEmail(email)) {
      dispatch(serError('Invalid email'));
      return false;
    }else if (password.length < 5) {
      dispatch(serError('Password should be at least 6 characters'));
      return false;
    }else if (password !== password2 || password.length < 5) {
      dispatch(serError('Passwords must match'));
      return false;
    }

    dispatch(removeError());
    return true;
  }

  return (
    <div>
      <h3 className="auth__title mb-5">RegisterScreen</h3>
      <form
        onSubmit={handleRegister}
        className="animate__animated animate__bounce animate__fadeIn animate__faster"
        >

        {
          msgError && ( <div className="auth__alert-error"> {msgError} </div> )
        }

        <input
          type="text"
          placeholder="Name"
          name="name"
          className="auth__input"
          autoComplete="off"
          value={name}
          onChange={handleInputChange}
        />
        <input
          type="text"
          placeholder="Email"
          name="email"
          className="auth__input"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
          autoComplete="off"
          value={password}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="Repeat Password"
          name="password2"
          className="auth__input"
          autoComplete="off"
          value={password2}
          onChange={handleInputChange}
        />
        <button
          className="btn btn-primary btn-block mb-5"
          type="submit"
        >
          Register
        </button>
        <Link className="link" to="/auth/login">
          Already register?
        </Link>
      </form>
    </div>
  );
};

export default RegisterScreen;
