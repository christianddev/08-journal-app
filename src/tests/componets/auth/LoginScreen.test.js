import React from 'react';
import { mount } from "enzyme"
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import LoginScreen from "../../../components/auth/LoginScreen"
import '@testing-library/jest-dom';
import { startLoginGoogle, startLoginWithEmailPassword } from '../../../actions/auth';
jest.mock('../../../actions/auth', () => ({
  startLoginGoogle: jest.fn(),
  startLoginWithEmailPassword: jest.fn()
}))

const middlewares = [thunk]
const mockStore = configureStore(middlewares)
const initState = {
  auth:{},
  ui:{
    loading: false,
    msgError: null
  }
}
let store = mockStore(initState)

store.dispatch = jest.fn()

const wrapper = mount (
  <Provider store={store}>
    <MemoryRouter>
      <LoginScreen />
    </MemoryRouter>
  </Provider>
)

describe('loginScreen.js', () => {

  beforeEach(() => {
    store = mockStore(initState)
    jest.clearAllMocks()
  })

  test('should show', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('should trigger startLoginGoogle', () => {
    wrapper.find('.google-btn').prop('onClick')()
    expect(startLoginGoogle).toHaveBeenCalled()
  })

  test('should no call startLoginWithEmailPassword', () => {
    wrapper.find('form').prop('onSubmit')({
      preventDefault(){}
    });
    expect(startLoginWithEmailPassword).not.toHaveBeenCalled()
  })
  test('should call startLoginWithEmailPassword', () => {

    const eventEmail = {
      target: {
        value: 'email@email.com',
        name: "email"
      }
    };
    wrapper.find("#inputEmail").simulate("change", eventEmail);
    const eventPassword = {
      target: {
        value: '123123',
        name: "password"
      }
    };
    wrapper.find("#inputPassword").simulate("change", eventPassword);

    wrapper.find('form').prop('onSubmit')({
      preventDefault(){}
    });
    expect(startLoginWithEmailPassword).toHaveBeenCalledWith(eventEmail.target.value, eventPassword.target.value)


  })


})
