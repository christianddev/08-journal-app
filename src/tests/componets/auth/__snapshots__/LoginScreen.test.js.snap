// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loginScreen.js should show 1`] = `
<div>
  <h3
    className="auth__title mb-5"
  >
    Login
  </h3>
  <form
    className="animate__animated animate__bounce animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      id="inputEmail"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      id="inputPassword"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block mb-5"
      disabled={false}
      type="submit"
    >
      Login
    </button>
    <hr />
    <div
      className="auth__social-networks"
    >
      <p>
        Login with Social Networks
      </p>
      <div
        className="google-btn"
        onClick={[Function]}
      >
        <div
          className="google-icon-wrapper"
        >
          <img
            alt="google button"
            className="google-icon"
            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
          />
        </div>
        <p
          className="btn-text"
        >
          <b>
            Sign in with google
          </b>
        </p>
      </div>
      <a
        className="link"
        href="/auth/register"
        onClick={[Function]}
      >
        Register
      </a>
    </div>
  </form>
</div>
`;
